File: pharmacy-api/local-api/dbsetup.go
package main

import (
	"log"
	"os"

	local_models "pharmacy-api/local-api/models"
	shared_models "pharmacy-api/shared/models"

	"golang.org/x/crypto/bcrypt"
	"gorm.io/driver/mysql"
	"gorm.io/gorm"
)

func setupDatabase() (db *gorm.DB) {
	dbUser := os.Getenv("DB_USER")
	dbPassword := os.Getenv("DB_PASSWORD")
	dbName := os.Getenv("DB_NAME")
	dbHost := os.Getenv("DB_HOST")
	dbPort := os.Getenv("DB_PORT")

	dsn := dbUser + ":" + dbPassword + "@tcp(" + dbHost + ":" + dbPort + ")/" + dbName + "?charset=utf8mb4&parseTime=True&loc=Local"

	db, err := gorm.Open(mysql.Open(dsn), &gorm.Config{})
	if err != nil {
		log.Fatalf("failed to connect to database: %v", err)
	}
	log.Println("Database connected")

	if err := db.AutoMigrate(
		&shared_models.WorkTime{},
		&shared_models.Producer{},
		&shared_models.Tag{},
		&shared_models.GoodsOrders{},
		&shared_models.Order{},
		&shared_models.Goods{},
		&shared_models.Announcement{},
		&local_models.User{},
		&local_models.Role{},
		&local_models.Permission{},
	); err != nil {
		log.Fatalf("failed to migrate database: %v", err)
	}

	// MYSQL or GORM moment idk, on Postgres all alright
	db.Exec("ALTER TABLE work_times MODIFY COLUMN time_start TIME")
	db.Exec("ALTER TABLE work_times MODIFY COLUMN time_end TIME")
	db.Exec("ALTER TABLE work_times MODIFY COLUMN date DATE")

	var persmissionsAmount int64
	db.Model(&local_models.Permission{}).Count(&persmissionsAmount)
	if persmissionsAmount == 0 {
		permissions := []local_models.Permission{
			{Action: "Update_Goods"},
			{Action: "Read_Orders"},
			{Action: "Update_Orders"},
			{Action: "Create_Orders"},
			{Action: "Delete_Orders"},
			{Action: "Update_WorkTime"},
			{Action: "Create_WorkTime"},
			{Action: "Delete_WorkTime"},
			{Action: "Update_Announces"},
			{Action: "Create_Announces"},
			{Action: "Delete_Announces"},
			{Action: "Change_Users"},
			{Action: "Change_Roles"},
			{Action: "Download_Logs"},
		}
		db.Create(&permissions)

	}

	return
}

func setupAdmin(db *gorm.DB) {
	pass, err := bcrypt.GenerateFromPassword([]byte(os.Getenv("CONTROL_PANEL_ADMIN_PASSWORD")), bcrypt.DefaultCost)

	if err != nil {
		log.Fatalf("failed to generate admin password: %v", err)
	}

	var permissions []local_models.Permission
	db.Find(&permissions)

	adminRole := local_models.Role{
		Name:        "Admin",
		Permissions: permissions,
	}

	user := local_models.User{
		Login:        os.Getenv("CONTROL_PANEL_ADMIN_LOGIN"),
		Role:         adminRole,
		UserName:     "Admin",
		PasswordHash: pass,
	}
	db.Create(&user)
}

File: pharmacy-api/src/shared/setup/db_setup.ts
import { DataSource } from 'typeorm';
import * as bcrypt from 'bcrypt';
import { Goods, User, Role, Permission } from '@pharmacy/src/shared/models/database_models';
import { testData } from '@pharmacy/src/shared/setup/db_data';

export async function setupDB(dataSource: DataSource): Promise<void> {
  const queryRunner = dataSource.createQueryRunner();
  try {
    await queryRunner.connect();
    
    let goodsCount: Goods[] = null
    try {
      goodsCount = await dataSource.getRepository(Goods).createQueryBuilder('goods').limit(1).getMany();
      if (goodsCount.length === 0) {
        await testData(dataSource);
        console.log("TEST DATA USED!");
      }
    }
    catch {
      await dataSource.runMigrations();
      await testData(dataSource);
      console.log("TEST DATA USED!");
    }

    const permissionsCount = await dataSource.getRepository(Permission).count();
    
    if (permissionsCount === 0) {
      const permissions: Partial<Permission>[] = [
        { action: "Update_Goods" },
        { action: "Read_Orders" },
        { action: "Update_Orders" },
        { action: "Create_Orders" },
        { action: "Delete_Orders" },
        { action: "Update_Schedule" },
        { action: "Create_Schedule" },
        { action: "Delete_Schedule" },
        { action: "Update_Announces" },
        { action: "Create_Announces" },
        { action: "Delete_Announces" },
        { action: "Change_Users" },
        { action: "Change_Roles" },
        { action: "Download_Logs" },
      ];
      await dataSource.getRepository(Permission).save(permissions);
      console.log("Permissions initialized");
    }

    const usersCount = await dataSource.getRepository(User).count();

    if (usersCount === 0 && process.env.CONTROL_PANEL_ADMIN_PASSWORD && process.env.CONTROL_PANEL_ADMIN_LOGIN) {
      await setupAdmin(dataSource);
    }

  } catch (error) {
    console.error("Failed to setup database:", error);
    throw error;
  } finally {
    await queryRunner.release();
  }
}

export async function setupAdmin(dataSource: DataSource): Promise<void> {
  const saltRounds = 10;

  const passwordHash = await bcrypt.hash(process.env.CONTROL_PANEL_ADMIN_PASSWORD, saltRounds);
  
  const permissions = await dataSource.getRepository(Permission).find();
  
  const adminRole = new Role();
  adminRole.name = "Admin";
  adminRole.permissions = permissions;
  
  const savedRole = await dataSource.getRepository(Role).save(adminRole);
  
  const user = new User();
  user.login = process.env.CONTROL_PANEL_ADMIN_LOGIN;
  user.userName = "Admin";
  user.passwordHash = passwordHash;
  user.roleId = savedRole.id;
  user.role = savedRole;
  
  await dataSource.getRepository(User).save(user);
  console.log("Admin user created successfully");
}